trieste/acquisition/function/function.py

```diff
@@ -581,6 +585,282 @@ def _update_expected_improvement_fn(
+        :raise tf.errors.InvalidArgumentError: If ``sample_size`` is not positive, or ``jitter`` is
+            negative.
+        """
+        tf.debugging.assert_positive(sample_size)
+        tf.debugging.assert_greater_equal(jitter, 0.0)
+
+        super().__init__()
+
+        self._sample_size = sample_size
+        self._jitter = jitter
+
+    def __repr__(self) -> str:
+        """"""
+        return f"MonteCarloExpectedImprovement({self._sample_size!r}, jitter={self._jitter!r})"
+
+    def prepare_acquisition_function(
+        self,
+        model: HasReparamSampler,
+        dataset: Optional[Dataset] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param model: The model over the specified ``dataset``. Must have output dimension [1].
+        :param dataset: The data from the observer. Cannot be empty.
+        :return: The estimated *expected improvement* acquisition function.
+        :raise ValueError (or InvalidArgumentError): If ``dataset`` is not populated, ``model``
+            does not have an output dimension of [1] or does not have a ``reparam_sample`` method.
+        """
+        if not isinstance(model, HasReparamSampler):
+            raise ValueError(
+                f"MonteCarloExpectedImprovement only supports models with a reparam_sampler method;"
+                f"received {model.__repr__()}"
+            )
+
+        sampler = model.reparam_sampler(self._sample_size)
+
+        tf.debugging.Assert(dataset is not None, [])
+        dataset = cast(Dataset, dataset)
+        tf.debugging.assert_positive(len(dataset), message="Dataset must be populated.")
+
+        samples_at_query_points = sampler.sample(dataset.query_points, jitter=self._jitter)
+        mean = tf.reduce_mean(samples_at_query_points, axis=0)
+
+        tf.debugging.assert_shapes(
+            [(mean, ["_", 1])], message="Expected model with output dimension [1]."
+        )
+
+        eta = tf.reduce_min(mean, axis=0)
+
+        return monte_carlo_expected_improvement(sampler, eta)
+
+    def update_acquisition_function(
+        self,
+        function: AcquisitionFunction,
+        model: HasReparamSampler,
+        dataset: Optional[Dataset] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param function: The acquisition function to update.
+        :param model: The model. Must have output dimension [1]. Unused here.
+        :param dataset: The data from the observer. Cannot be empty
+        """
+        tf.debugging.Assert(dataset is not None, [])
+        dataset = cast(Dataset, dataset)
+        tf.debugging.assert_positive(len(dataset), message="Dataset must be populated.")
+        tf.debugging.Assert(isinstance(function, monte_carlo_expected_improvement), [])
+        sampler = function._sampler  # type: ignore
+        sampler.reset_sampler()
+        samples_at_query_points = sampler.sample(dataset.query_points, jitter=self._jitter)
+        mean = tf.reduce_mean(samples_at_query_points, axis=0)
```

Comment:
do you know how come in `BatchMonteCarloExpectedImprovement` predict call is used for computing the mean?

Reply:
I think `BatchMonteCarloExpectedImprovement` assumes that the mean can be computed analytically. Here it would probably be dangerous to do that, since for e.g. a DGP the `predict` call will return a sample which wouldn't even necessarily be smooth.